### PR TITLE
[3.7] bpo-27987: align PyGC_Head to alignof(long double)

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue27987

### CODE REVIEW

File: Include/objimpl.h [hunk 1]
```diff
@@ -255,7 +255,7 @@ typedef union _gc_head {
         union _gc_head *gc_prev;
         Py_ssize_t gc_refs;
     } gc;
-    double dummy;  /* force worst-case alignment */
+    long double dummy;  /* force worst-case alignment (16 byte for amd64) */
```
Comment (vstinner) [inline Include/objimpl.h]: Can you please elaborate the comment? Add "bpo-27987" and explain that x64-64 ABI requires to align on 16 bytes.

Do you *have to* align to 16 bytes on x86 (32-bit) as well?

Comment (methane) [reply Include/objimpl.h]: > Can you please elaborate the comment? Add "bpo-27987" and explain that x64-64 ABI requires to align on 16 bytes.

I'm not sure this is really x86-64 ABI requirement...
C99 requires malloc returns memory aligned to be used for any built-in types.

> Do you have to align to 16 bytes on x86 (32-bit) as well?

 I don't think so.  long double may be aligned to 8 byte boundary.
But I am not sure.  I don't have 32bit development environment for now.

**Comment by vstinner** [comment]
C code:

```
union {
    struct {
        void *next;
        void *prev;
    } _gc;
    long double alignment;
} PyGC_Head;

int main()
{
    return sizeof(PyGC_Head);
}
```

With long double:

```
vstinner@apu$ gcc x.c -o x && ./x; echo $? # 64-bit void*
16
vstinner@apu$ gcc -m32 x.c -o x && ./x; echo $? # 32-bit void*
12
```

Without long double:

```
vstinner@apu$ gcc x.c -o x && ./x; echo $? # 64-bit void*
16
vstinner@apu$ gcc -m32 x.c -o x && ./x; echo $? # 32-bit void*
8
```

I'm not sure that 12 bytes is great for alignment. 32-bit CPU may be more efficient on SIMD instructions with 8 bytes alignment, no?

**Comment by vstinner** [comment]
@methane also proposed PR #13336 for master, but he rejected it.

**Comment by methane** [comment]
PyGC_Head in master branch is two uintptr_t.  It is 8 bytes (not 12 bytes) on x86.

**Comment by methane** [comment]
> I'm not sure that 12 bytes is great for alignment. 32-bit CPU may be more efficient on SIMD instructions with 8 bytes alignment, no?

This pull request changes **from** 12 bytes to 12 or 16 bytes, no?

**Comment by vstinner** [comment]
Sorry @methane, I don't understand this change :-( What is the effect of PyListObject for example? Is the address of ob_item aligned to 16 bytes on x86 (32-bit)? Is it on x86-64 (64 bit)?

**Comment by methane** [comment]
@vstinner You wrote:

```c
    struct {
        void *next;
        void *prev;
    } _gc;
```

But it is actually:

```c
    struct {
        void *next;
        void *prev;
        ssize_t gc_refs; /* only in Python < 3.8 */
    } _gc;
```

This is very big difference between Python 3.7 and 3.8.

sizeof(PyGC_Head) is 12 byte on x86, regardless alignment is double or long double.  (I can now "gcc -m32" on Ubuntu, thank you)
On amd64, sizeof(PyGC_Head) is 24 byte with double alignment, and 32byte with long double alignment.

> What is the effect of PyListObject for example? Is the address of ob_item aligned to 16 bytes on x86 (32-bit)? Is it on x86-64 (64 bit)?

This pull request affects only where `PyObject*` (`(PyObject*)(gc + 1)`) is aligned.
This pull request doesn't affect alignment in PyListObject.  `offsetof(PyListObject, ob_item)` is 24.

**Review by vstinner — APPROVED** [review]
LGTM. I think that I understood the issue and that the fix is correct :-)

**Comment by gpshead** [comment]
this change needs to be undone.  followup on the bug.

**Comment by methane** [comment]
Do you think missing "build" directory is happened by this change?

I thought it just a trouble on the build bot environment?

**Comment by gpshead** [comment]
The resulting python interpreter was crashing on the the PGO bot and a pile of new undefined behavior was happening on the undefined behavior sanitizer (ubsan) bot.

PGO bot:
```
./python -E -S -m sysconfig --generate-posix-vars ;\
if test $? -ne 0 ; then \
	echo "generate-posix-vars failed" ; \
	rm -f ./pybuilddir.txt ; \
	exit 1 ; \
fi
Segmentation fault
generate-posix-vars failed
Makefile:605: recipe for target 'pybuilddir.txt' failed
```

undefined behavior sanitizer millions of times:
```
Objects/funcobject.c:663:5: runtime error: member access within misaligned address 0x7fb4b5a03a08 for type 'union _gc_head', which requires 16 byte alignment
0x7fb4b5a03a08: note: pointer points here
 b4 7f 00 00  70 8d c6 b5 b4 7f 00 00  a0 13 a0 b5 b4 7f 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^ 
Objects/funcobject.c:663:5: runtime error: member access within misaligned address 0x7fb4b5a03a08 for type 'struct (anonymous struct at ./Include/objimpl.h:253:5)', which requires 16 byte alignment
0x7fb4b5a03a08: note: pointer points here
 b4 7f 00 00  70 8d c6 b5 b4 7f 00 00  a0 13 a0 b5 b4 7f 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^ 
Objects/funcobject.c:663:5: runtime error: store to misaligned address 0x7fb4b5a03a08 for type 'union _gc_head *', which requires 16 byte alignment
0x7fb4b5a03a08: note: pointer points here
 b4 7f 00 00  70 8d c6 b5 b4 7f 00 00  a0 13 a0 b5 b4 7f 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^ 
```

**Comment by methane** [comment]
#13318 fixed it already.

**Pull request merged**